Revit_Core_Engine/Query/UniqueParameterNames.cs

```diff
@@ -0,0 +1,179 @@
+            {
+                if (elementsByDoc.ContainsKey(element.Document))
+                    elementsByDoc[element.Document].Add(element);
+                else
+                    elementsByDoc.Add(element.Document, new List<Element> { element });
+            }
+
+            return elementsByDoc;
+        }
+
+        /***************************************************/
+
+        [Description("Creates dictionary for elements per category.")]
+        private static Dictionary<ElementId, List<Element>> ElementsByCategoryDictionary(this IEnumerable<Element> elements)
+        {
+            Dictionary<ElementId, List<Element>> elementsByCat = new Dictionary<ElementId, List<Element>>();
+
+            foreach (Element element in elements)
+            {
+                if (elementsByCat.ContainsKey(element.Category.Id))
+                    elementsByCat[element.Category.Id].Add(element);
+                else
+                    elementsByCat.Add(element.Category.Id, new List<Element> { element });
+            }
+
+            return elementsByCat;
+        }
+
+        /***************************************************/
+
+        [Description("Returns unique parameter ids for the collection of the elements.")]
+        private static HashSet<int> UniqueParametersIds(this IEnumerable<Element> elementsFromOneDocument)
+        {
+            HashSet<int> ids = new HashSet<int>();
+            
+            foreach (Element element in elementsFromOneDocument)
+            {
+                foreach (Parameter parameter in element.ParametersMap)
+                {
+                    ids.Add(parameter.Id.IntegerValue);
+                }
+            }
+
+            return ids;
+        }
+
+        /***************************************************/
+
+        [Description("Returns unique element types for the collection of the elements.")]
+        private static HashSet<int> UniqueTypeIds(this IEnumerable<Element> elementsFromOneDocument)
+        {
+            HashSet<int> ids = new HashSet<int>();
+            Document doc = elementsFromOneDocument.FirstOrDefault().Document;
+
+            foreach (Element element in elementsFromOneDocument)
+            {
+                Element elementType = doc.GetElement(element.GetTypeId());
+
+                if (elementType != null)
+                    ids.Add(elementType.Id.IntegerValue);
+            }
+
+            return ids;
+        }
+
+        /***************************************************/
+
+        [Description("Returns name of the paramter based on id and document it belongs.")]
+        private static string ParameterName(this Document doc, int id)
+        {
+            string parameterName;
+
+            if (id < 0)
+                parameterName = LabelUtils.GetLabelFor((BuiltInParameter)id);
+            else
+                parameterName = (doc.GetElement(new ElementId(id)) as ParameterElement).GetDefinition().Name;
```

Comment:
Curious if `ParameterElement` does not already have the name we are looking for?